libyul/optimiser/StackLimitEvader.cpp

```diff
@@ -0,0 +1,142 @@
+
+	uint64_t run(YulString _function = YulString{})
+	{
+		if (nextAvailableSlot.count(_function))
+			return nextAvailableSlot[_function];
+
+		// Assign to zero early to guard against recursive calls.
+		nextAvailableSlot[_function] = 0;
+
+		uint64_t nextSlot = 0;
+		if (callGraph.count(_function))
+			for (YulString child: callGraph.at(_function))
+				nextSlot = std::max(run(child), nextSlot);
+
+		if (unreachableVariables.count(_function))
+		{
+			yulAssert(!slotAllocations.count(_function), "");
+			auto& assignedSlots = slotAllocations[_function];
+			for (YulString variable: unreachableVariables.at(_function))
+				if (variable.empty())
+				{
+					// TODO: Too many function arguments or return parameters.
+				}
+				else
+					assignedSlots[variable] = nextSlot++;
+		}
+
+		return (nextAvailableSlot[_function] = nextSlot);
+	}
+
+	map<YulString, map<YulString, uint64_t>> slotAllocations{};
+	map<YulString, uint64_t> nextAvailableSlot{};
+};
+u256 getLiteralArgumentValue(FunctionCall const& _call)
+{
+	yulAssert(_call.arguments.size() == 1, "");
+	Literal const* literal = std::get_if<Literal>(&_call.arguments.front());
+	yulAssert(literal, "");
+	return valueOfLiteral(*literal);
+}
+}
+
+void StackLimitEvader::run(
+	OptimiserStepContext& _context,
+	Object& _object,
+	map<YulString, set<YulString>> const& _unreachableVariables
+)
+{
+	yulAssert(_object.code, "");
+	auto const* evmDialect = dynamic_cast<EVMDialect const*>(&_context.dialect);
+	yulAssert(
+		evmDialect && evmDialect->providesObjectAccess(),
+		"StackToMemoryMover can only be run on objects using the EVMDialect with object access."
+	);
+
+	vector<FunctionCall*> getFreeMemoryStartCalls = FunctionCallFinder::run(
+		*_object.code,
+		"memoryguard"_yulstring
+	);
+	if (getFreeMemoryStartCalls.empty())
+		return;
+
+	// Make sure all calls to ``memoryguard`` we found have the same value as argument.
+	u256 reservedMemory = getLiteralArgumentValue(*getFreeMemoryStartCalls.front());
+	for (FunctionCall const* getFreeMemoryStartCall: getFreeMemoryStartCalls)
+			if (reservedMemory != getLiteralArgumentValue(*getFreeMemoryStartCall))
+				return;
+
+	CallGraph callGraph = CallGraphGenerator::callGraph(*_object.code);
+
+	// We cannot move variables in recursive functions to fixed memory offsets.
+	for (YulString function: callGraph.recursiveFunctions())
+		if (_unreachableVariables.count(function))
+			return;
+
+	MemoryOffsetAllocator memoryOffsetAllocator{_unreachableVariables, callGraph.functionCalls};
+	uint64_t requiredSlots = memoryOffsetAllocator.run();
+
+	StackToMemoryMover{_context, reservedMemory, memoryOffsetAllocator.slotAllocations}(*_object.code);
+	reservedMemory += 32 * requiredSlots;
+	for (FunctionCall* getFreeMemoryStartCall: getFreeMemoryStartCalls)
+	{
+		Literal* literal = std::get_if<Literal>(&getFreeMemoryStartCall->arguments.front());
+		literal->value = YulString{util::toCompactHexWithPrefix(reservedMemory)};
```

Comment:
Add assert for `literal`.

Reply:
No real need since it was already checked above (via ``getLiteralArgumentValue``)... but I can assert it again.

Reply:
We're asserting at dereferencing to avoid random crashes. And the code above this may change and then this part here is left naked.